### ISSUE TITLE
Hardcode public IP in peer

### ISSUE THREAD

**Issue by mgoelzer** [libp2p/go-libp2p#460]
@lucasvo and I are discussing a scenario.  The basic question is:  can you force a node to think its public IP is x.y.z.a even if that's not an interface on the machine?

Scenario:  user has two libp2p nodes, both of which are in docker containers on AWS behind ELBs such that the machines (containers) themselves don't know what their public IP is.  But the admin knows, and it's a fixed IP number.  Do we provide a way to force a node to operate with a specific IP hard coded by the programmer?

(By the way, wouldn't this issue come up in IPFS as well?  There must be people running IPFS behind ELBs.)

cc @vyzo @Stebalien

**Comment by vyzo** [comment]
Yes, you can advertise the address through the address factory.

**Comment by lucasvo** [comment]
@vyzo, going by this: https://godoc.org/github.com/libp2p/go-libp2p/p2p/host/basic#AddrsFactory what we would do is create a method that takes any input and replaces the IPv4 part of the multiaddr with a preconfigured IP?

**Comment by Stebalien** [comment]
Yes. Or, if the port is stable, you can just return a per-defined address (that's what go-ipfs does).

**Comment by Stebalien** [comment]
Yes.

**Comment by Stebalien** [comment]
Actually, a slightly better way would be to iterate over each addr in `addrs`, replacing `/ipX/.../` with the external IP.

**Comment by mikiquantum** [comment]
Ah, I get it. AddressFactory will be used to only broadcast the addresses that the node is available at. 
If let's say we want to broadcast a node A with the internal private ip (local private network with peers) and the external ip (of a load balancer) for external connections. 
How does the resolver work from a peer B attempting to open a connection to the peer A, as the peerstore for peer A will have 2 entries (1 private ip, 1 public)? 
Which multi-address will chose first? It will always select the public one over the private one? or random order?
If it is random, I guess there will be a time penalty on choosing the `wrong` one.

Sorry if this is very clear in the code base.
Thanks!

**Comment by vyzo** [comment]
This is the subject of ongoing refactoring (the dialing logic -- see https://github.com/libp2p/go-libp2p-swarm/pull/88).
It will currently dial both addresses concurrently, with the faster one "winning".

**Comment by raulk** [comment]
@mikiquantum right now there is no specific order, but we're making the dialer more intelligent here: https://github.com/libp2p/go-libp2p-swarm/pull/88. 

Not only will you be able to plug in the "dial planning strategy" that's most appropriate for your app, but we'll ship with a more intelligent planner that uses advertised IPs to determine which dials to prioritise, e.g. if the peer and you share the same public IP, we'll prioritise private IPs.

**Comment by mikiquantum** [comment]
Thanks for the info @vyzo and @raulk , I will keep an eye on it! 
For now as @Stebalien suggested, probably hardcoding the externalIP is the best choice.

Thank you!
